### PR TITLE
Implement len caching for parking_lot RwLock

### PR DESCRIPTION
**Pull request by mattrutherford**
- Rename crate to `len-caching-lock`
- Implement for `parking_lot::RwLock`
- Implement `Len` trait for all Rust `std::collections`
- Improve docs and tests

Required for a different ongoing [WIP PR](https://github.com/paritytech/parity-ethereum/pull/9987)

Closes [#10026](https://github.com/paritytech/parity-ethereum/issues/10026)

### CODE REVIEW

File: util/len-caching-lock/src/mutex.rs [hunk 1]
```diff
@@ -13,30 +13,15 @@
...
-impl<T> Len for Vec<T> {
-	fn len(&self) -> usize { self.len() }
-}
-
-impl<T> Len for VecDeque<T> {
-	fn len(&self) -> usize { self.len() }
-}
-
-/// Can be used in place of a `Mutex` where reading `T`'s `len()` without 
+use Len;
```
Comment (andresilva) [inline util/len-caching-lock/src/mutex.rs]: newline?

File: util/len-caching-lock/src/mutex.rs [hunk 2]
```diff
@@ -46,62 +31,78 @@ pub struct LenCachingMutex<T> {
...
 			len: &self.len,
 		}
 	}
 
-	pub fn try_lock(&self) -> Option<Guard<T>> {
-		Some( Guard {
+	/// Delegates to `parking_lot::Mutex`
+	/// [`try_lock()`](../../lock_api/struct.Mutex.html#method.try_lock).
+	pub fn try_lock(&self) -> Option<CachingMutexGuard<T>> {
+		Some( CachingMutexGuard {
```
Comment (andresilva) [inline util/len-caching-lock/src/mutex.rs]:
```suggestion
		Some(CachingMutexGuard {
```

File: util/len-caching-lock/src/rwlock.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use Len;
+
+/// Can be used in place of a [`RwLock`](../../lock_api/struct.RwLock.html) where 
+/// reading `T`'s `len()` without needing to lock, is advantageous. 
+/// When the WriteGuard is released, `T`'s `len()` will be cached.
+pub struct LenCachingRwLock<T> {
+  data: RwLock<T>,
```
Comment (andresilva) [inline util/len-caching-lock/src/rwlock.rs]: use tabs instead.

File: util/len-caching-lock/src/rwlock.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		CachingRwLockWriteGuard {
+			write_guard: self.data.write(),
+			len: &self.len,
+		}
+	}
+
+	/// Delegates to `parking_lot::RwLock`
+	/// [`try_write()`](../../lock_api/struct.RwLock.html#method.try_write).
+	pub fn try_write(&self) -> Option<CachingRwLockWriteGuard<T>> {
+		Some( CachingRwLockWriteGuard {
```
Comment (andresilva) [inline util/len-caching-lock/src/rwlock.rs]:
```suggestion
		Some(CachingRwLockWriteGuard {
```

File: util/len-caching-lock/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+/// Implement to allow a type with a len() method to be used
+/// with [`LenCachingMutex`](mutex/struct.LenCachingMutex.html)
+/// or  [`LenCachingRwLock`](rwlock/struct.LenCachingRwLock.html)
+pub trait Len {
+	fn len(&self) -> usize;
+}
+
+impl<T> Len for Vec<T> {
+	fn len(&self) -> usize { self.len() }
```
Comment (tomusdrw) [inline util/len-caching-lock/src/lib.rs]: Isn't that recursive? Shouldn't it be more like `Vec::len(self)`?

File: util/len-caching-lock/src/mutex.rs [hunk 6]
```diff
@@ -13,30 +13,15 @@
...
-impl<T> Len for VecDeque<T> {
-	fn len(&self) -> usize { self.len() }
-}
-
-/// Can be used in place of a `Mutex` where reading `T`'s `len()` without 
+use Len;
+/// Can be used in place of a [`Mutex`](../../lock_api/struct.Mutex.html) where reading `T`'s `len()` without 
 /// needing to lock, is advantageous. 
 /// When the Guard is released, `T`'s `len()` will be cached.
 /// The cached `len()` may be at most 1 lock behind current state.
```
Comment (tomusdrw) [inline util/len-caching-lock/src/mutex.rs]: Should have a `Default` and non-blocking `Debug` implementation.

File: util/len-caching-lock/src/mutex.rs [hunk 7]
```diff
@@ -46,62 +31,78 @@ pub struct LenCachingMutex<T> {
...
+	/// Load the cached value that was returned from your `T`'s `len()`
+	/// subsequent to the most recent lock being released.
 	pub fn load_len(&self) -> usize {
 		self.len.load(Ordering::SeqCst)
 	}
 
-	pub fn lock(&self) -> Guard<T> {
-		Guard {
+	/// Convenience method to check if collection T `is_empty()`
+	pub fn load_is_empty(&self) -> bool {
```
Comment (tomusdrw) [inline util/len-caching-lock/src/mutex.rs]: I don't think it's worthwhile unless it's specialised via another `AtomicBool` (which probably doesn't make much sense). I'd remove that method, doing `self.load_len() == 0` is self-explanatory.

File: util/len-caching-lock/src/rwlock.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+use std::sync::atomic::Ordering;
+
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use Len;
+
+/// Can be used in place of a [`RwLock`](../../lock_api/struct.RwLock.html) where 
+/// reading `T`'s `len()` without needing to lock, is advantageous. 
+/// When the WriteGuard is released, `T`'s `len()` will be cached.
+pub struct LenCachingRwLock<T> {
```
Comment (tomusdrw) [inline util/len-caching-lock/src/rwlock.rs]: Same as with `Mutex`, please add `Default` and `Debug`

File: util/len-caching-lock/src/rwlock.rs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  data: RwLock<T>,
+  len: AtomicUsize,
+}
+
+impl<T: Len> LenCachingRwLock<T> {
```
Comment (tomusdrw) [inline util/len-caching-lock/src/rwlock.rs]: Might be good to relax to `?Sized` bound here

**Comment by mattrutherford** [comment]
Thanks @andresilva - now done.

File: util/len-caching-lock/src/lib.rs (hunk 5 above)
Comment (mattrutherford) [reply util/len-caching-lock/src/lib.rs]: I thought struct methods always had higher priority than trait, but perhaps it's better to be explicit.

File: util/len-caching-lock/src/mutex.rs (hunk 7 above)
Comment (mattrutherford) [reply util/len-caching-lock/src/mutex.rs]: Sure, I was debating whether to provide it...

File: util/len-caching-lock/src/mutex.rs (hunk 6 above)
Comment (mattrutherford) [reply util/len-caching-lock/src/mutex.rs]: Also added `From` implementation

File: util/len-caching-lock/src/rwlock.rs (hunk 8 above)
Comment (mattrutherford) [reply util/len-caching-lock/src/rwlock.rs]: Done

**Comment by mattrutherford** [comment]
Thanks for the review and comments @tomusdrw !

**Pull request merged**